spark/src/main/scala/org/apache/spark/sql/delta/MetadataCleanup.scala

```diff
@@ -126,6 +154,172 @@ trait MetadataCleanup extends DeltaLogging {
+    if (availableNonV2Checkpoints.nonEmpty) {
+      metrics.v2CheckpointCompatLogicTimeTakenMs = System.currentTimeMillis() - startTimeMs
+      return
+    }
+
+    // topLevelFileIndex must be non-empty when topLevelFiles are present
+    val shallowCopyDf =
+      loadIndex(snapshotToCleanup.checkpointProvider.topLevelFileIndex.get, Action.logSchema)
+    val finalPath =
+      FileNames.checkpointFileSingular(snapshotToCleanup.deltaLog.logPath, checkpointVersion)
+    Checkpoints.createCheckpointV2ParquetFile(
+      spark,
+      shallowCopyDf,
+      finalPath,
+      hadoopConf,
+      useRename = false)
+    metrics.v2CheckpointCompatLogicTimeTakenMs = System.currentTimeMillis() - startTimeMs
+    metrics.checkpointVersion = checkpointVersion
+  }
+
+  /** Deletes any unreferenced files from the sidecar directory `_delta_log/_sidecar` */
+  protected def identifyAndDeleteUnreferencedSidecarFiles(
+      snapshotToCleanup: Snapshot,
+      checkpointRetention: Long,
+      metrics: SidecarDeletionMetrics): Unit = {
+    val startTimeMs = System.currentTimeMillis()
+    // If v2 checkpoints are not enabled on the table, we don't need to attempt the sidecar cleanup.
+    if (!CheckpointProvider.isV2CheckpointEnabled(snapshotToCleanup)) return
+
+    val hadoopConf = newDeltaHadoopConf()
+    val fs = sidecarDirPath.getFileSystem(hadoopConf)
+    // This can happen when the V2 Checkpoint feature is present in the Protocol but
+    // only Classic checkpoints have been created for the table.
+    if (!fs.exists(sidecarDirPath)) return
+
+    val (parquetCheckpointFiles, otherFiles) = store
+      .listFrom(listingPrefix(logPath, 0), hadoopConf)
+      .collect { case CheckpointFile(status, _) => (status, CheckpointInstance(status.getPath)) }
+      .collect { case (fileStatus, ci) if ci.format.usesSidecars => fileStatus }
+      .toSeq
+      .partition(_.getPath.getName.endsWith("parquet"))
+    val (jsonCheckpointFiles, unknownFormatCheckpointFiles) =
+      otherFiles.partition(_.getPath.getName.endsWith("json"))
+    if (unknownFormatCheckpointFiles.nonEmpty) {
+      logWarning(
+        "Found checkpoint files other than parquet and json: " +
+          s"${unknownFormatCheckpointFiles.map(_.getPath.toString).mkString(",")}")
+    }
+    metrics.numActiveParquetCheckpointFiles = parquetCheckpointFiles.size
+    metrics.numActiveJsonCheckpointFiles = jsonCheckpointFiles.size
+    val parquetCheckpointsFileIndex =
+      DeltaLogFileIndex(DeltaLogFileIndex.CHECKPOINT_FILE_FORMAT_PARQUET, parquetCheckpointFiles)
+    val jsonCheckpointsFileIndex =
+      DeltaLogFileIndex(DeltaLogFileIndex.CHECKPOINT_FILE_FORMAT_JSON, jsonCheckpointFiles)
+    val identifyActiveSidecarsStartTimeMs = System.currentTimeMillis()
+    metrics.activeCheckpointsListingTimeTakenMs = identifyActiveSidecarsStartTimeMs - startTimeMs
+    val sparkSession = spark
+    // scalastyle:off sparkimplicits
+    import sparkSession.implicits._
```

Comment:
Why needed? The error message explains why it's bad, and also how to import Delta-optimized implicits